Libraries/LibGfx/JPGLoader.cpp

```diff
@@ -0,0 +1,1057 @@
+            if (!read_reset_marker(stream, context))
+                return false;
+            break;
+        case JPG_DHT:
+            if (!read_huffman_table(stream, context))
+                return false;
+            break;
+        case JPG_SOS:
+            return read_start_of_scan(stream, context);
+        default:
+            if (!skip_marker_with_length(stream)) {
+                dbg() << stream.offset() << String::format(": Error skipping marker: %x!", marker);
+                return false;
+            }
+            break;
+        }
+    }
+
+    ASSERT_NOT_REACHED();
+}
+
+static bool scan_huffman_stream(BufferStream& stream, JPGLoadingContext& context)
+{
+    u8 last_byte;
+    u8 current_byte;
+    stream >> current_byte;
+
+    for (;;) {
+        last_byte = current_byte;
+        stream >> current_byte;
+        if (stream.handle_read_failure()) {
+            dbg() << stream.offset() << ": EOI not found!";
+            return false;
+        }
+
+        if (last_byte == 0xFF) {
+            if (current_byte == 0xFF)
+                continue;
+            if (current_byte == 0x00) {
+                stream >> current_byte;
+                context.huffman_stream.stream.append(last_byte);
+                continue;
+            }
+            Marker marker = 0xFF00 | current_byte;
+            if (marker == JPG_EOI)
+                return true;
+            if (marker >= JPG_RST0 && marker <= JPG_RST7) {
+                context.huffman_stream.stream.append(marker);
+                stream >> current_byte;
+                continue;
+            }
+            dbg() << stream.offset() << String::format(": Invalid marker: %x!", marker);
+            return false;
+        } else {
+            context.huffman_stream.stream.append(last_byte);
+        }
+    }
+
+    ASSERT_NOT_REACHED();
+}
+
+static bool load_jpg_impl(JPGLoadingContext& context)
+{
+    ByteBuffer buffer = ByteBuffer::wrap(context.compressed_data, context.compressed_size);
+    BufferStream stream(buffer);
+    if (!parse_header(stream, context))
+        return false;
+    if (!scan_huffman_stream(stream, context))
+        return false;
+
+    auto result = decode_huffman_stream(context);
+    if (!result.has_value()) {
+        dbg() << stream.offset() << ": Failed to decode Macroblocks!";
+        return false;
+    } else {
+        auto macroblocks = result.release_value();
+        dbg() << String::format("%i macroblocks decoded successfully :^)", macroblocks.size());
+        dequantize(context, macroblocks);
+        inverse_dct(context, macroblocks);
+        ycbcr_to_rgb(context, macroblocks);
+        compose_bitmap(context, macroblocks);
+    }
+    return true;
+}
+
+JPGImageDecoderPlugin::JPGImageDecoderPlugin(const u8* data, size_t size)
+{
+    m_context = make<JPGLoadingContext>();
+    m_context->compressed_data = data;
+    m_context->compressed_size = size;
+    m_context->huffman_stream.stream.ensure_capacity(500 * KB);
```

Comment:
500 Kb of preallocated buffer seems a little... excessive. But I don't know enough about JPEG to know if it's a size that could be reached often.
Reading the code, though, it feels like it should never really get bigger than the file itself so, in that case, it could at least be sized with `min(500 * KB, size)` instead of always allocating a flat 500 Kb.
Is this something that could / should be more properly sized ?

Reply:
You're right. 500kb is too much. I'm not sure what I was thinking there.